modules/stochastic_tools/src/covariances/ExponentialCovariance.C

```diff
@@ -40,54 +41,57 @@ ExponentialCovariance::ExponentialCovariance(const InputParameters & parameters)
 }
 
 void
-ExponentialCovariance::computeCovarianceMatrix(RealEigenMatrix & K,
-                                               const RealEigenMatrix & x,
-                                               const RealEigenMatrix & xp,
+ExponentialCovariance::computeCovarianceMatrix(torch::Tensor & K,
+                                               const torch::Tensor & x,
+                                               const torch::Tensor & xp,
                                                const bool is_self_covariance) const
 {
-  if ((unsigned)x.cols() != _length_factor.size())
+  if ((unsigned)x.sizes()[1] != _length_factor.size())
     mooseError("length_factor size does not match dimension of trainer input.");
 
   ExponentialFunction(
       K, x, xp, _length_factor, _sigma_f_squared, _sigma_n_squared, _gamma, is_self_covariance);
 }
 
 void
-ExponentialCovariance::ExponentialFunction(RealEigenMatrix & K,
-                                           const RealEigenMatrix & x,
-                                           const RealEigenMatrix & xp,
+ExponentialCovariance::ExponentialFunction(torch::Tensor & K,
+                                           const torch::Tensor & x,
+                                           const torch::Tensor & xp,
                                            const std::vector<Real> & length_factor,
                                            const Real sigma_f_squared,
                                            const Real sigma_n_squared,
                                            const Real gamma,
                                            const bool is_self_covariance)
 {
-  unsigned int num_samples_x = x.rows();
-  unsigned int num_samples_xp = xp.rows();
-  unsigned int num_params_x = x.cols();
+  unsigned int num_samples_x = x.sizes()[0];
+  unsigned int num_samples_xp = xp.sizes()[0];
+  unsigned int num_params_x = x.sizes()[1];
 
-  mooseAssert(num_params_x == xp.cols(),
+  mooseAssert(num_params_x == xp.sizes()[1],
               "Number of parameters do not match in covariance kernel calculation");
 
+  auto K_accessor = K.accessor<Real, 2>();
+  auto x_accessor = x.accessor<Real, 2>();
+  auto xp_accessor = xp.accessor<Real, 2>();
   for (unsigned int ii = 0; ii < num_samples_x; ++ii)
   {
     for (unsigned int jj = 0; jj < num_samples_xp; ++jj)
     {
       // Compute distance per parameter, scaled by length factor
       Real r_scaled = 0;
       for (unsigned int kk = 0; kk < num_params_x; ++kk)
-        r_scaled += pow((x(ii, kk) - xp(jj, kk)) / length_factor[kk], 2);
+        r_scaled += pow((x_accessor[ii][kk] - xp_accessor[jj][kk]) / length_factor[kk], 2);
       r_scaled = sqrt(r_scaled);
-      K(ii, jj) = sigma_f_squared * std::exp(-pow(r_scaled, gamma));
+      K_accessor[ii][jj] = sigma_f_squared * std::exp(-pow(r_scaled, gamma));
     }
     if (is_self_covariance)
-      K(ii, ii) += sigma_n_squared;
+      K_accessor[ii][ii] += sigma_n_squared;
   }
 }
```

Comment:
This can be a potential cause of slowdown. Libtorch has optimized routines to compute these matrix operations.